### PR TITLE
 Mark "canProgress" when unfinished operator outputted null or empty page

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 1]
```diff
@@ -377,18 +377,14 @@ private void processNewSources()
                     // get an output page from current operator
                     current.getOperatorContext().startIntervalTimer();
                     Page page = current.getOutput();
+                    movedPage = true;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Rename, since it changed its meaning.

**Comment by sopel39** [comment]
> We need to make sure we don't enable Driver to loop now while doing nothing useful

That's contract between Driver and operators. How can you check that?

> This might be still insufficient. I didn't think about this previously, but Operator#finish()`` can also make progress: see HBO.finish()` is where, by necessity, all the spill magic happens.

How does it break the contract?

**Comment by sopel39** [comment]
ac. I need https://github.com/prestodb/presto/pull/10004 too. Still don't know how HBO `finish` breaks something. Does it yield?

**Comment by sopel39** [comment]
I need to add tests to `TestDriver` too

**Comment by dain** [comment]
I'm going to assume this is still being discussed in the issue.

**Comment by sopel39** [comment]
@dain This is an improvement over existing code and does not require driver interface to be changed, but prerequisite is: https://github.com/prestodb/presto/pull/10004

**Pull request closed without merging**